Add SOP for worker restart (#13)
3scale/3scale-Operations: sops/alerts/apicast_worker_restart.adoc

:toc:
:toc-placement!:

= ThreescaleApicastWorkerRestart

toc::[]

== Description

> * This alert will trigger if one of the nginx-worker restarted for any reason.

Reviewer: ```suggestion
* This alert will trigger if one of the nginx-worker restarted for any reason
```

== Troubleshooting

* Check why the worker died in the Apicast logs; the usual reasons are:

>   - Pod hits the memory limits, and Cgroups kill one of the worker processes.

Reviewer: ```suggestion
- Pod hits the memory limits and one of the worker processes was killed. Check why the limits were reached, and if it's needed, increase them
```

>   - Node is out of memory, so try to kill containers to schedule to another node.

Reviewer: ```suggestion
- Node is out of memory, so try to kill containers to schedule to another node
```

> == Verification

Reviewer: This section should have information on how to verify that the issue is solved.

Author: How do you verify that a limit has reached and it's working ok? sending request to the pod?

Reviewer: In the Pod events you can find what limit was reached and the reasons of the Pod restart, for example:

```
Events:
  Type     Reason     Age                From                                 Message
  ----     ------     ----               ----                                 -------
  Warning  Unhealthy  19m (x5 over 20m)  kubelet, ip-10-65-7-70.ec2.internal  Liveness probe failed: Get http://10.131.6.168:9394/metrics: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
  Normal   Killing    19m                kubelet, ip-10-65-7-70.ec2.internal  Container system-sidekiq failed liveness probe, will be restarted
```

You can verify the pod is working if you see that is running and handling requests. Also you can also check if the pod consumption is estable or rising by using the `Top` command or looking the Prometheus / Grafana dashboards. It the logs are OK and the memory/cpu is stable, you can consider the alert resolved.

```
❯ k top pods -n 3scale backend-worker-584f87f695-2gg7c
NAME                              CPU(cores)   MEMORY(bytes)   
backend-worker-584f87f695-2gg7c   0m           56Mi    
```

Author: The thing is, the pod does not restart, this is why we added this metric.  :-(

Reviewer: In that case, I think the verification will be that this metrics stabilizes and remains constant (so the workers are no longer being restarted). What do you think?

Author: PR updated.

> * If the kill happens because pod limits, check why the limits were reached,
>   and if it's needed, increase the memory.

Reviewer: ```suggestion
```

This is not actually a verification, moved to the troubleshooting section.

> * Suppose node killed the container, nothing to do. This behaviour will be
>   changed in the next releases.

Reviewer: ```suggestion
```

This also need to go to the troubleshooting section and if you can add a bit more context so why a node may kill the container, it would be great.
